contracts/protocol/clients/voucher/BosonVoucher.sol

```diff
@@ -279,4 +496,143 @@ contract BosonVoucher is IBosonVoucher, BeaconClientBase, OwnableUpgradeable, ER
     }
+
+    /**
+     * @notice Performs special case transfer hooks.
+     *
+     * - Commits to pre-minted offers on first transfer.
+     *
+     * When a voucher is pre-minted, the owner is not stored
+     * but the seller is reported as the transferee and owner,
+     * so that it can be found by marketplaces and listed by
+     * the seller for their asking price.
+     *
+     * On the first transfer after pre-minting, this method
+     * will commit to the associated offer on behalf of the
+     * new owner.
+     *
+     * - Updates buyer on subsequent transfers.
+     *
+     * When a voucher with an associated exchange is transferred
+     * either on the secondary market or just between wallets,
+     * the protocol needs to be alerted to the change of buyer
+     * address.
+     *
+     * The buyer account associated with the exchange will be
+     * replaced. If the new voucher holder already has a
+     * Boson Protocol buyer account, it will be used. Otherwise,
+     * a new buyer account will be created and associated with
+     * the exchange.
+     *
+     * @param from - the address from which the voucher is being transferred
+     * @param to - the address to which the voucher is being transferred
+     * @param tokenId - the tokenId of the voucher that is being transferred
+     */
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override {
+        // Update the buyer associated with the voucher in the protocol
+        // Only when transferring, not minting or burning
+        if (from == owner()) {
+            if (premintStatus.committable) {
+                // If this is a transfer of premited token, treat it differently
+                address protocolDiamond = IClientExternalAddresses(BeaconClientLib._beacon()).getProtocolAddress();
+                // TODO: uncomment when commitToPreMintedOffer method is available
+                // IBosonExchangeHandler(protocolDiamond).commitToPreMintedOffer(to, premintStatus.offerId, tokenId);
+                delete premintStatus;
+            } else {
+                // Already committed, treat as a normal transfer
+                onVoucherTransferred(tokenId, payable(to));
+            }
+        } else if (from != address(0) && to != address(0) && from != to) {
+            onVoucherTransferred(tokenId, payable(to));
+        }
+    }
+
+    /**
+     * @dev Determines if a token is pre-minted and committable via transfer hook
+     *
+     * Committable means:
+     * - does not yet have an owner
+     * - in a reserved range
+     * - has been pre-minted
+     * - has not been already burned
+     *
+     * @param _tokenId - the token id to check
+     * @return committable - whether the token is committable
+     * @return offerId - the associated offer id if committable
+     */
+    function getPreMintStatus(uint256 _tokenId) internal view returns (bool committable, uint256 offerId) {
```

Comment:
I wonder whether or not we would need to do this binary search if we stored another HashMap, that included "used pre-minted voucher" or something similar in concept?

I am just putting this thought down, not attempting to block progress on this work, just questioning whether or not we really want to be performing a binary search here ?

Reply:
I guess this implies that querying for a token that hasn't been committed to when we do have pre-mints available, is always going to be costly ? Where would we be doing this in normal operation in the protocol ?

Reply:
In subsequent PR #485 a hashmap for used tokens is already implemented. 
But that doesn't mean that we can go without binary search. 

Binary search happens in 
- `ownerOf` where it doesn't matter how costly it is, since it's used only externally
- during the first transfer of a pre-minted voucher. In any other valid transfer binary search never really happens.